components/button/index.js

```diff
@@ -25,28 +29,60 @@ export function Button( props, ref ) {
 		className,
 		disabled,
 		focus,
+		children,
 		...additionalProps
 	} = props;
 
+	const tag = href !== undefined && ! disabled ? 'a' : 'button';
+	const tagProps = tag === 'a' ? { href, target, rel } : { type: 'button', disabled };
+	const { icon = true } = additionalProps;
+	const isExternalLink = href && ( target && ! includes( [ '_self', '_parent', '_top' ], target ) );
+
 	const classes = classnames( 'components-button', className, {
 		button: ( isPrimary || isLarge || isSmall ),
 		'button-primary': isPrimary,
 		'button-large': isLarge,
 		'button-small': isSmall,
 		'is-toggled': isToggled,
 		'is-busy': isBusy,
+		'external-link': isExternalLink,
+		'components-icon-button': isExternalLink && icon && ( isPrimary || isLarge || isSmall ),
 	} );
 
-	const tag = href !== undefined && ! disabled ? 'a' : 'button';
-	const tagProps = tag === 'a' ? { href, target } : { type: 'button', disabled };
+	const getRel = () => {
+		// Allow to omit the `rel` attribute passing a `null` value.
+		return rel === null ? rel : uniq( compact( [
+			...rel.split( ' ' ),
+			'external',
+			'noreferrer',
+			'noopener',
+		] ) ).join( ' ' );
+	};
+
+	const { opensInNewTabText = __(
+		/* translators: accessibility text */
+		'(opens in a new tab)'
+	) } = additionalProps;
+
+	const OpensInNewTabScreenReaderText = isExternalLink && (
+		<span className="screen-reader-text">
+			{
+				// We need a space to separate this from previous text.
+				' ' + opensInNewTabText
+			}
+		</span>
+	);
+
+	const ExternalIcon = icon && isExternalLink && <Dashicon icon="external" />;
```

Comment:
This will behave funny if you assign `icon={ 0 }`, since the behavior of JS in an `&&` assignment is to defer to the left value if it is falsey, i.e.

```js
const ExternalIcon = 0 && true && <Dashicon icon="external" />;
```

... will evaluate to:

```js
const ExternalIcon = 0;
```

And React happily outputs a literal `0` text ([Example](https://codepen.io/aduth/pen/dWwzrL))

(Only `false`, `""`, `null`, or an `undefined` child will not be rendered)

Unless you know that all members of the condition are boolean values, I'd suggest a ternary instead, where the fallback value is one of the above unrendered (most likely `null`).